[META] Contributing Guide Revamp (#24705)
facebook/react-native: .github/wiki/Tests.md

# Tests

## Running Tests
This section is about testing your changes to React Native as a contributor. If you haven't yet, go through the steps to set up your development environment for [building projects with native code][env-setup]. If you are intested in [writing tests](#writing-tests) for a React Native app, proceed to the next section.

[env-setup]: http://facebook.github.io/react-native/docs/getting-started

### JavaScript Tests

The simplest way to run the JavaScript test suite is by using the following command at the root of your React Native checkout:

```bash

> npm test

Reviewer: ```suggestion
yarn test
```

Reviewer: Actually, just noticed that yarn is missing from the setup instructions as well

Author: Yarn is not used in the Getting Started guide intentionally, as it was written around the same time Yarn was released and the additional steps to install it were deemed extraneous as npm is already bundled with node.

I think we can update the website to use yarn, at which point we can do the same here. After all, we use yarn in our CI over npm.

Reviewer: Yeah that makes sense - the only risk is that new devs may run into some strange dependency issues with the difference in how deps are installed between yarn and npm 👍

```

This will run tests using [Jest](https://jestjs.io).

You should also make sure your code passes [Flow](https://flowtype.org/) and lint tests:

```bash
npm run flow
npm run lint
```

### iOS Tests

To run the iOS tests, invoke the following script from the root of your React Native checkout:

```bash

> ./scripts/objc-test-ios.sh test

Reviewer: Should these have npm scripts?

Author: They don't at this time, but we should set it up. There's a lot of cleanup that can be done here, for example, there's three files: objc-test-ios.sh, objc-test-tvos.sh, and objc-test.sh. The latter should not be invoked directly, as the first two just set up the necessary envvars.

I actually wrote most of this doc over three months ago but have deferred adding it as I kept encountering areas where the DX could be improved, so I am leaning towards documenting how things are today even if its painful in some areas. Getting more eyes on it may help us get an idea which of these we should focus on fixing first.

Reviewer: Ah yeah totally agree

Once you've done that, you can start the Android emulator using:

```bash
./scripts/run-android-emulator.sh
```

Then, run the Android integration tests:

```bash
./scripts/run-android-local-integration-tests.sh
```

### End-to-end Tests

Finally, make sure end-to-end tests run successfully by executing the following script:

```bash
./scripts/test-manual-e2e.sh
```

End-to-end tests written in [Detox](https://github.com/wix/Detox) confirm that React Native components and APIs function correctly in the context of a running app. They run the RNTester app in the simulator and simulate a user interacting with the app.

You can run Detox end-to-end tests locally by [installing the Detox CLI](https://github.com/wix/Detox/blob/master/docs/Introduction.GettingStarted.md#step-1-install-dependencies) on macOS, then running the following in the command line:

```bash
npm run build-ios-e2e
npm run test-ios-e2e
```

If you work on a component or API that isn't convered by a Detox test, please consider adding one. Detox tests are stored under [`RNTester/e2e/__tests__`](https://github.com/facebook/react-native/tree/master/RNTester/e2e/__tests__).

> ### Continuous Testing

Reviewer: This section is great, should it be mentioned in the contributing doc?

We use [Appveyor][config-appveyor] and [Circle CI][config-circleci] to automatically run our open source tests. Appveyor and Circle CI will run these tests whenever a commit is added to a pull request, as a way to help maintainers understand whether a code change introduces a regression. The tests also run on commits to the master and `*-stable` branches in order to keep track of the health of these branches.

[config-appveyor]: https://github.com/facebook/react-native/blob/master/.appveyor/config.yml
[config-circleci]: https://github.com/facebook/react-native/blob/master/.circleci/config.yml

> There's another set of tests that run within Facebook's internal test infrastructure. Some of these tests are integration tests defined by internal consumers of React Native (e.g. unit tests for a React Native surface in the Facebook app). These tests run on every commit to the copy of React Native hosted on Facebook's source control. They also run when a pull request is imported to Facebook's source control. As it happens, should one of these tests fail, you'll need someone at Facebook to take a look. Since pull requests can only be imported by Facebook employees, whoever imported the pull request should be able to facilitate any details.

Reviewer: ```suggestion
There's another set of tests that run within Facebook's internal test infrastructure. Some of these tests are integration tests defined by internal consumers of React Native (e.g. unit tests for a React Native surface in the Facebook app). These tests run on every commit to the copy of React Native hosted on Facebook's source control. They also run when a pull request is imported to Facebook's source control.

If one of these tests fail, you'll need someone at Facebook to take a look. Since pull requests can only be imported by Facebook employees, whoever imported the pull request should be able to facilitate any details.
```
